### PR TITLE
Introduce fs and dynamodb locks for publish

### PR DESCRIPTION
**Pull request by agadelshin**

### CODE REVIEW

**Comment by TimSimpson** [comment]
In addition to locking the body of the publish methods like this we need to make the code aware of S3's eventual read consistency issues by storing the latest ETag from S3.

The ["ETag"](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html) is an MD5 digest of the object data received when writing changes to S3. This gets returned when writing data to S3 or doing a read.

Currently there is a possibility that a user may write data to S3, then a second user could update the package later but be given an older version of the object by S3, even if the code was gated with locking.

I have not worked with Azure blob storage or GCS but my guess is they both have the same issue with eventual consistency which would cause older data to be returned to users who would then unknowingly destroy previous updates when they wrote their changes at the end of the publish method.

One fix is to make Locker's `Lock()` method return the ETag or equivalent for the given uri. So the code in `publisher_s3.go` could use it and make sure the data it's grabbing from S3 matches, and also store this info when it uploads new data. 

I'm just spitballing here, but the code would look something like this:

```go
import "crypto/md5"

...

func (s *S3Publisher) publish(osarch *zps.OsArch, pkgFiles []string, zpkgs []*zps.Pkg, keyPair *KeyPairEntry) error {

...
    // Lock changes to return some kind of metadata for each publisher. For s3 it would be the etag:
    eTag, err = locker.Lock()
    if err != nil {
        return fmt.Errorf("repository: %s is locked by another process, error: %s", s.name, err.Error())
    }

    defer locker.Unlock()

    var metadataDb *osFile
    
    defer metadataDb.Close()

    // Download metadata db
    client := s3manager.NewDownloader(s.session)

    retries := 5

    for {
        metadataDb, err := os.Create(metaPath)
        if err != nil {
            return err
        }
        _, err = client.Download(metadataDb, &s3.GetObjectInput{
            Bucket: aws.String(s.uri.Host),
            Key:    aws.String(path.Join(s.uri.Path, osarch.String(), "metadata.db")),
        })
        if err != nil {
            return errors.New(fmt.Sprintf("unable to download: %s", s.uri.Path))
        }
        bytes, err := ioutil.ReadFile(metadataDb)
        if err != nil {
            return fmt.Errorf("Unable to read download S3 object: %s", s.uri.Path)
        }
        actualETag := md5.Sum(bytes)
        if actualETag != eTag {
            retries -= 1
            if retries > 0 {
                metadataDb.Close()
                continue;
            }
            return fmt.Errorf("S3 object %q has eTag mismatch: want %q, got %q", s.uri.Path, eTag, actualETag)
        }
    }
    defer metadataDb.Close()

    ...

    bytes, err := ioutil.ReadFile(configSig)
    if err != nil {
        return fmt.Errorf("Unable to read repackaged file for upload: %s", configSig)
    }
    newETag := md5.Sum(bytes)
    
    // the Unlock method would need to accept the new ETag somehow
    locker.Unlock(newETag)
    _, err = uploader.Upload(&s3manager.UploadInput{
        Bucket: aws.String(s.uri.Host),
        Key:    aws.String(path.Join(s.uri.Path, "config.sig")),
        Body:   configSig,
    })
```

Of course, the dynamo locker implementation would need to change as well.

In the past when I've designed Dynamo tables to keep track of concurrently updated S3 objects like this, the key is typically the S3 key path (like it is here), and then there's an Attribute for the Etag and another one which is a timestamp for when the Item was last updated and finally a boolean attribute for whether or not the Item is locked. When problems arose where locks were inadvertently kept the date made it easy to know if we could manually remove the locks or not.

File: zpm/publisher_s3.go [hunk 1]
```diff
@@ -338,12 +343,35 @@ func (s *S3Publisher) channel(osarch *zps.OsArch, pkg string, channel string, ke
...
+	for {
+		size, err := client.Download(metadataDb, &s3.GetObjectInput{
+			Bucket: aws.String(s.uri.Host),
+			Key:    aws.String(path.Join(s.uri.Path, osarch.String(), "metadata.db")),
+		})
+		if err != nil {
+			return fmt.Errorf("unable to download: %s", s.uri.Path)
+		}
+
+		medatabaDbData := make([]byte, size)
```
Comment (TimSimpson) [inline zpm/publisher_s3.go]: Should this be `metadataDbData`?

File: zpm/lock_dynamo.go [hunk 2]
```diff
@@ -32,25 +33,58 @@ func NewDynamoLocker(uri *url.URL) *DynamoLocker {
...
 func (d *DynamoLocker) Lock() error {
-	_, err := d.client.PutItem(&dynamodb.PutItemInput{
-		Item: map[string]*dynamodb.AttributeValue{
+	_, err := d.LockWithEtag()
+	return err
+}
+
+func (d *DynamoLocker) Unlock() error {
+	eTag := [16]byte{}
+	return d.UnlockWithEtag(eTag)
```
Comment (TimSimpson) [inline zpm/lock_dynamo.go]: This needs to update dynamo but leave the ETag as it is. It can then be used in situations where the lock is given up but the ETag is not changed (such as when there is some kind of error updating the resource).

Additionally to make it safe to always call this with defer, the lock should remember if either version of `unlock` has been called yet by storing a boolean called `Locked` which will be set when Lock is called and unset when either `Unlock` is called. 

So it should look something like this:

```go

func (d *DynamoLocker) Unlock() error {
	if !d.Locked {
		return nil
	}

	_, err := d.client.UpdateItem(&dynamodb.UpdateItemInput{
		Key: map[string]*dynamodb.AttributeValue{
			"LockKey": {S: aws.String(d.key)},
		},
		UpdateExpression: aws.String("SET Locked = :false"),
		ExpressionAttributeValues: map[string]*dynamodb.AttributeValue{
			":false": {BOOL: aws.Bool(false)},
		},
		TableName: &d.tableName,
	})

	if err != nil {
		return err
	}
	d.locked = false
	return nil
}
```

File: zpm/publisher_abs.go [hunk 3]
```diff
@@ -397,20 +435,42 @@ func (a *ABSPublisher) publish(osarch *zps.OsArch, pkgFiles []string, zpkgs []*z
...
 	locker := NewLocker(a.lockUri)
 
-	err = locker.Lock()
+	eTag, err := locker.LockWithEtag()
 	if err != nil {
 		return fmt.Errorf("repository: %s is locked by another process, error: %s", a.name, err.Error())
 	}
 
-	defer locker.Unlock()
+	defer locker.UnlockWithEtag(newEtag)
```
Comment (TimSimpson) [inline zpm/publisher_abs.go]: This needs to be `defer locker.Unlock()`, and `defer.UnlockWithEtag(newTag)` should be called with the happy path at the bottom, once the new etag has been created.

This applies to all code using defer for unlocking.

File: zpm/publisher_abs.go [hunk 4]
```diff
@@ -485,6 +545,14 @@ func (a *ABSPublisher) publish(osarch *zps.OsArch, pkgFiles []string, zpkgs []*z
 			return err
 		}
 
+		medatabaDbData := make([]byte, 0)
+		_, err = metadataUp.Read(medatabaDbData)
+		if err != nil {
+			return fmt.Errorf("unable to read downloaded file: %s", a.uri.Path)
+		}
+
+		eTag = md5.Sum(medatabaDbData)
```
Comment (TimSimpson) [inline zpm/publisher_abs.go]: Somewhere below, after the upload succeeds the new eTag should be updated by calling `locker.UnlockWithEtag`

This applies to the other publishers as well.

File: zpm/publisher_s3.go [hunk 5]
```diff
@@ -521,6 +585,14 @@ func (s *S3Publisher) publish(osarch *zps.OsArch, pkgFiles []string, zpkgs []*zp
 			Body:   metadataUp,
 		})
 
+		medatabaDbData := make([]byte, 0)
+		_, err = metadataDb.Read(medatabaDbData)
+		if err != nil {
+			return fmt.Errorf("unable to read downloaded file: %s", s.uri.Path)
+		}
+
+		eTag = md5.Sum(medatabaDbData)
```
Comment (TimSimpson) [inline zpm/publisher_s3.go]: I believe there is a way to pass the etag to S3 in the request when uploading a file. This ensures that the file's MD5 sum doesn't change due to network glitches.

File: zpm/lock_dynamo.go [hunk 6]
```diff
@@ -32,25 +33,58 @@ func NewDynamoLocker(uri *url.URL) *DynamoLocker {
...
+}
+
+func (d *DynamoLocker) LockWithEtag() ([16]byte, error) {
```
Comment (TimSimpson) [inline zpm/lock_dynamo.go]: If this gets used for multiple publishers, since ETag seems a little S3 specific maybe we should call it `LockWithMd5Sum`?

**Review by TimSimpson — CHANGES_REQUESTED** [review]

File: zpm/lock_dynamo.go (hunk 6 above)
Comment (TimSimpson) [reply zpm/lock_dynamo.go]: Nevermind, I mispoke, GCS and ABS both call it etag too, it's not an AWS concept.

**Comment by TimSimpson** [comment]
That said I think the current approach is solid and should work. However here's an idea in the future.

Azure provides mechanisms for ensuring consistency amid concurrent access. 
https://docs.microsoft.com/en-us/azure/storage/blobs/concurrency-manage?tabs=dotnet

Google actually does something similar - see here under `Preventing the race condition`:
https://cloud.google.com/storage/docs/generations-preconditions

So - an option for the future is to couple the Dynamo table with s3 access, so locking happens by default for all s3_publishers. Then use the ability to send the etag you get on initial read of ABS or GCS objects when writing it to make sure it hasn't changed and concurrent access can be fixed there as well.

File: zpm/publisher_abs.go (hunk 4 above)
Comment (agadelshin) [reply zpm/publisher_abs.go]: it's called in defer now

```
defer locker.UnlockWithEtag(&eTag)
```

I do believe it should be fine since it's a pointer

File: zpm/publisher_s3.go (hunk 1 above)
Comment (agadelshin) [reply zpm/publisher_s3.go]: thanks :)

File: zpm/lock_dynamo.go (hunk 6 above)
Comment (agadelshin) [reply zpm/lock_dynamo.go]: yeah, terminology came from http specification

File: zpm/publisher_s3.go (hunk 5 above)
Comment (agadelshin) [reply zpm/publisher_s3.go]: I almost sure it should be on lower levels, I see in newer version of sdk there is ETag field from UploadOutput https://github.com/aws/aws-sdk-go/blob/main/service/s3/s3manager/upload.go#L100-L115, maybe it makes sense to update sdk for zps, but it might be challenging :)

File: zpm/lock_dynamo.go (hunk 2 above)
Comment (agadelshin) [reply zpm/lock_dynamo.go]: I took another approach and call `d.UnlockWithEtag(&eTag)`, so eTag is passed by reference. After metadata is uploaded, we change eTag value and defer will be called with new value

https://play.golang.org/p/Rvk2HksphoT - tested it in playground

File: zpm/publisher_abs.go (hunk 3 above)
Comment (agadelshin) [reply zpm/publisher_abs.go]: see comment above

File: zpm/lock_dynamo.go (hunk 2 above)
Comment (TimSimpson) [reply zpm/lock_dynamo.go]: There are three places in the code where you want to remove the lock, but not update the etag. Having a method that explicitly does that is in my opinion is clearer and avoids needing to check the calling code is correct in those spots. I agree though that the approach you've taken should work.

File: zpm/publisher_s3.go (hunk 5 above)
Comment (TimSimpson) [reply zpm/publisher_s3.go]: Ok, I opened [an issue](https://github.com/fezz-io/zps/issues/28) so it can be addressed later.

File: Makefile [hunk 7]
```diff
@@ -30,13 +30,13 @@ $(os):
 	mkdir -p -m 0750 dist/${@}-x86_64/var/cache/zps
 	mkdir -p -m 0750 dist/${@}-x86_64/var/tmp/zps
 	mkdir -p -m 0755 dist/${@}-x86_64/usr/share/zps/certs/zps.io
-	cp ../zps.io/ca.pem ../zps.io/intermediate.pem ../zps.io/zps.pem dist/${@}-x86_64/usr/share/zps/certs/zps.io
+	#cp ../zps.io/ca.pem ../zps.io/intermediate.pem ../zps.io/zps.pem dist/${@}-x86_64/usr/share/zps/certs/zps.io
```
Comment (TimSimpson) [inline Makefile]: I'm curious - why was this changed? It seems unrelated.

Comment (agadelshin) [reply Makefile]: changed it for local development, reverted! thanks

**Review by TimSimpson — APPROVED** [review]
